Makefile

```diff
@@ -1,59 +1,149 @@
+checkfmt: SHELL := /usr/bin/env bash
+checkfmt: ## Check formatting of all go files
+	@ $(MAKE) --no-print-directory log-$@
+	@ echo "$(RESULT)"
+	@ if [ "$(RESULT)" != "OK" ]; then exit 1; fi
 
-build-windows-amd64:
-	GOOS=windows GOARCH=amd64 $(GOBUILD) -o bin/windows-amd64/$(NAME).exe
+.PHONY: test
+test: ## Run tests
+	@ $(MAKE) --no-print-directory log-$@
+	$(GOCMD) test $(MODVENDOR) -v $(GOPKGS)
 
-.PHONY: clean
-clean:
-	rm -rf ./bin
+###################
+## Build targets ##
+###################
+.PHONY: build
+build: clean ## Build binary for current OS/ARCH
+	@ $(MAKE) --no-print-directory log-$@
+	$(GOBUILD) -o ./$(BUILD_DIR)/$(GOOS)-$(GOARCH)/$(NAME)
+
+.PHONY: build-all
+build-all: GOOS      = linux darwin windows freebsd
+build-all: GOARCH    = amd64 arm
+build-all: clean ## Build binary for all OS/ARCH
+	@ $(MAKE) --no-print-directory log-$@
+	@ ./hack/build/build-all-osarch.sh "$(BUILD_DIR)" "$(NAME)" "$(VERSION)" "$(GOOS)" "$(GOARCH)" $(GOLDFLAGS)
+
+#####################
+## Release targets ##
+#####################
+.PHONY: release patch minor major
+PATTERN =
+
+release: version ?= $(shell echo $(VERSION) | sed 's/^v//' | awk -F'[ .]' '{print $(PATTERN)}')
+release: ## Prepare release
+	@ $(MAKE) --no-print-directory log-$@
+	@ ./hack/release/release.sh "$(version)" "$(VERSION)"
+
+patch: PATTERN = '\$$1\".\"\$$2\".\"\$$3+1'
+patch: release ## Prepare Patch release
+
+minor: PATTERN = '\$$1\".\"\$$2+1\".0\"'
+minor: release ## Prepare Minor release
+
+major: PATTERN = '\$$1+1\".0.0\"'
+major: release ## Prepare Major release
+
+####################
+## Helper targets ##
+####################
+.PHONY: authors
+authors: ## Generate Authors
+	git log --all --format='%aN <%aE>' | sort -u | egrep -v noreply > AUTHORS
 
 .PHONY: changelog
-changelog:
+changelog: ## Generate Changelog
 	git-chglog -o CHANGELOG.md
 
-.PHONY: deps
-deps:
-	GO111MODULE=on go mod vendor
-
-.PHONY: lint
-lint:
-	golangci-lint run ./...
-
-.PHONY: release
-release:
-	git tag -a v$(VERSION) -m v$(VERSION)
-	git push --tags
-
-.PHONY: test
-test:
-	go test -v $(GOPKGS)
+.PHONY: goimports
+goimports: ## Install goimports
+	@ $(MAKE) --no-print-directory log-$@
+	GO111MODULE=off go get -u golang.org/x/tools/cmd/goimports
+
+.PHONY: tools
+tools: ## Install required tools
+	@ $(MAKE) --no-print-directory log-$@
+	curl -sfL https://install.goreleaser.com/github.com/golangci/golangci-lint.sh | sh -s  -- -b $(shell go env GOPATH)/bin $(GOLANGCI_VERSION)
+	curl -sfL https://github.com/git-chglog/git-chglog/releases/download/$(GITCHGLOG_VERSION)/git-chglog_$(shell go env GOOS)_$(shell go env GOARCH) -o $(shell go env GOPATH)/bin/git-chglog && chmod +x $(shell go env GOPATH)/bin/git-chglog
+	GO111MODULE=off go get -u github.com/mitchellh/gox
+
+####################################
+## Self-Documenting Makefile Help ##
+####################################
+.PHONY: help
+help:
```

Comment:
Appears as magical. Here, and also in the log target, some documentation would really be beneficial.

Reply:
+1